File: file_merger.py
import csv
import argparse

class FileMerger:
    def __init__(self, files, key_columns, separators):
        self.files = files
        self.key_columns = key_columns
        self.separators = separators
        self.data = {}

    def read_data(self):
        """
        Reads the data from each file into a dictionary, using the key column as the key
        """
        for file, separator, key_column in zip(self.files, self.separators, self.key_columns):
            with open(file, 'r') as f:
                reader = csv.reader(f, delimiter=separator)
                self.data[file] = {row[key_column]: row for row in reader}

    def merge_data(self):
        """
        Merges the data from the dictionaries and returns the merged rows
        """
        keys = set.intersection(*[set(d.keys()) for d in self.data.values()])
        merged_rows = []
        for key in keys:
            merged_row = []
            for file, separator, key_column in zip(self.files, self.separators, self.key_columns):
                row = self.data[file].get(key)
                if row is None:
                    row = ["-" for _ in range(len(separator))]
                merged_row.extend(row)
            merged_rows.append(merged_row)
        return merged_rows

    def write_to_file(self, output_file):
        """
        Writes the merged rows to the specified output file
        """
        with open(output_file, 'w') as f:
            writer = csv.writer(f)
            for row in self.merge_data():
                writer.writerow(row)

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('files', nargs='+', help='The files to merge')
    parser.add_argument('--key-columns', nargs='+', type=int, help='The key column(s) in each file')
    parser.add_argument('--separators', nargs='+', help='The separator(s) for each file')
    parser.add_argument('--output-file', required=True, help='The output file')
    args = parser.parse_args()

    merger = FileMerger(args.files, args.key_columns, args.separators)
    merger.read_data()
    merger.write_to_file(args.output_file)

File: file_merger.cpp
#include <iostream>
#include <fstream>
#include <unordered_map>
#include <set>
#include <vector>

class FileMerger {
public:
    FileMerger(const std::vector<std::string>& files, const std::vector<int>& key_columns, const std::vector<char>& separators)
        : files_(files), key_columns_(key_columns), separators_(separators) {}

    void ReadData() {
        data_.clear();
        for (size_t i = 0; i < files_.size(); ++i) {
            std::ifstream file(files_[i]);
            std::string line;
            while (std::getline(file, line)) {
                std::vector<std::string> row;
                size_t start = 0;
                size_t end = line.find(separators_[i]);
                while (end != std::string::npos) {
                    row.emplace_back(line.substr(start, end - start));
                    start = end + 1;
                    end = line.find(separators_[i], start);
                }
                row.emplace_back(line.substr(start));
                data_[files_[i]][row[key_columns_[i]]] = row;
            }
        }
    }

    std::vector<std::vector<std::string>> MergeData() {
        std::set<std::string> keys;
        for (const auto& [file, rows] : data_) {
            std::set<std::string> file_keys;
            for (const auto& [key, row] : rows) {
                file_keys.insert(key);
            }
            if (keys.empty()) {
                keys = std::move(file_keys);
            } else {
                std::set<std::string> new_keys;
                std::set_intersection(keys.begin(), keys.end(), file_keys.begin(), file_keys.end(),
                                      std::inserter(new_keys, new_keys.begin()));
                keys = std::move(new_keys);
            }
        }

        std::vector<std::vector<std::string>> merged_rows;
        for (const auto& key : keys) {
            std::vector<std::string> merged_row;
            for (size_t i = 0; i < files_.size(); ++i) {
                const auto& rows = data_[files_[i]];
                const auto it = rows.find(key);
                if (it != rows.end()) {
                    merged_row.insert(merged_row.end(), it->second.begin(), it->second.end());
                } else {
                    merged_row.emplace_back("-");
                }
            }
            merged_rows.emplace_back(std::move(merged_row));
        }
        return merged_rows;
    }

    void WriteToFile(const std::string& output_file) {
        std::ofstream file(output_file);
       
